src/it/scala/io/iohk/ethereum/sync/FastSyncItSpec.scala

```diff
@@ -0,0 +1,461 @@
+                        (updateWorldForBlock: (BigInt, InMemoryWorldStateProxy) => InMemoryWorldStateProxy): (Block, BigInt, InMemoryWorldStateProxy) = {
+      val newBlockNumber = parent.header.number + 1
+      val newWorld = updateWorldForBlock(newBlockNumber, parentWorld)
+      val newBlock = parent.copy(header = parent.header.copy(parentHash = parent.header.hash, number = newBlockNumber, stateRoot = newWorld.stateRootHash))
+      val newTd = newBlock.header.difficulty + parentTd
+      (newBlock, newTd, parentWorld)
+    }
+
+    def importNBlocksToTheTopForm(startState: BlockchainState, n: Int)
+                                 (updateWorldForBlock: (BigInt, InMemoryWorldStateProxy) => InMemoryWorldStateProxy): Task[Unit] = {
+      def go(parent: Block, parentTd: BigInt, parentWorld: InMemoryWorldStateProxy, blocksLeft: Int): Task[Unit] = {
+        if (blocksLeft <= 0) {
+          Task.now(())
+        } else {
+          val (newBlock, newTd, newWorld) = createChildBlock(parent, parentTd, parentWorld)(updateWorldForBlock)
+          bl.save(newBlock, Seq(), newTd, saveAsBestBlock = true)
+          bl.persistCachedNodes()
+          go(newBlock, newTd, newWorld, blocksLeft - 1)
+        }
+      }
+
+      go(startState.bestBlock, startState.currentTd, startState.currentWorldState, n)
+    }
+
+    def syncUntil(n: BigInt)(updateWorldForBlock: (BigInt, InMemoryWorldStateProxy) => InMemoryWorldStateProxy): Task[Unit] = {
+      Task(bl.getBestBlock()).flatMap { block =>
+        if (block.number >= n) {
+          Task(())
+        } else {
+          Task {
+            val currentTd = bl.getTotalDifficultyByHash(block.hash).get
+            val currentWolrd = getMptForBlock(block)
+            val (newBlock, newTd, newWorld) = createChildBlock(block, currentTd, currentWolrd)(updateWorldForBlock)
+            bl.save(newBlock, Seq(), newTd, saveAsBestBlock = true)
+            bl.persistCachedNodes()
+            broadcastBlock(newBlock, newTd)
+          }.flatMap(_ => syncUntil(n)(updateWorldForBlock))
+        }
+      }
+    }
+
+    def startFastSync(): Task[Unit] = Task {
+      fastSync ! FastSync.Start
+    }
+
+    def waitForFastSyncFinish(): Task[Boolean] = {
+      retryUntilWithDelay(Task(storagesInstance.storages.appStateStorage.isFastSyncDone()), 1.second, 90){ isDone =>
+        isDone
+      }
+    }
+
+    // Reads whole trie into memory, if the trie lacks nodes in storage it will be None
+    def getBestBlockTrie(): Option[MptNode] = {
+      Try {
+        val bestBlock = bl.getBestBlock()
+        val bestStateRoot = bestBlock.header.stateRoot
+        MptTraversals.parseTrieIntoMemory(HashNode(bestStateRoot.toArray), storagesInstance.storages.stateStorage.getBackingStorage(bestBlock.number))
+      }.toOption
+    }
+  }
+
+  object FakePeer {
+    def startFakePeer(peerName: String): Task[FakePeer] = {
+      for {
+        peer <- Task(new FakePeer(peerName)).memoizeOnSuccess
```

Comment:
It is `def` so `.memoizeOnSuccess` won't work here

Reply:
some leftover from debugging i will remove it